File: interpr-scala/src/main/scala/zeppelin/Interpreter.scala
package zeppelin

import org.apache.thrift._
import org.apache.thrift.protocol._
import org.apache.thrift.server._
import org.apache.thrift.transport._
import org.apache.thrift.server.TServer.Args

class InterpreterServerImpl extends InterpreterServer.Iface {
  override def shutdown() = {
    println("Sutting down Scala interpreter")
    //TODO suicide, exit 0
    println("Done.")  
  }

  override def interprete(text: String): InterpreterResult = {
    println("Interpreting: " + text)
    new InterpreterResult("Ok")
  }
}

object Interpreter extends App {
  val port = 9090
  val host = "localhost"

  println("Starting Zeppelin interpreter")
  try {
    val server = buildServer()    
    server.serve();     
  } catch {
    case x: Exception => x.printStackTrace();
  }

  def buildServer() = {
    println("Listnenig port " + port)
    val transport = new TServerSocket(port)
    val processor = new InterpreterServer.Processor(new InterpreterServerImpl())

    //val server = new TThreadPoolServer(new Args(transport).processor(processor))
    val server = new TSimpleServer(new Args(transport)
      .processor(processor)
      .protocolFactory(new TCompactProtocol.Factory())
      .transportFactory(new TFramedTransport.Factory())
    )
    server
  }
}

File: interpr-python/interpreter.py
#!/usr/bin/python
#Pure python interpreter for Zeppelin

#Intended to be run as a separate process and communicate with
# Z though Thrift RPC

import sys
import socket
sys.path.append('./gen-py')

from zeppelin import InterpreterServer
from zeppelin.ttypes import *

from thrift import Thrift
from thrift.transport import TSocket
from thrift.transport import TTransport
from thrift.protocol import TCompactProtocol



class InterpreterServerHandler:
  def __init__(self):
    self.log = {}

  def sutdown(self):
    print "Shuting Interpreter down"

  def interprete(self, text):
    print "Interpreting: ", text
    return InterpreterResult("Done by " + socket.gethostbyname(socket.gethostname()))


def main():
  try:
    handler = InterpreterServerHandler()
    processor = InterpreterServer.Processor(handler)
    transport = TSocket.TServerSocket(9090)
    tfactory = TTransport.TFramedTransportFactory()
    pfactory = TCompactProtocol.TCompactProtocolFactory()
     
    server = TServer.TSimpleServer(processor, transport, tfactory, pfactory)
     
    print "Starting Python Interpreter server..."
    server.serve()
    print "Done."
  
  except Thrift.TException, tx:
    print '%s' % (tx.message)

if __name__ == '__main__':
  main()
